### PR TITLE
Various LCD / backlight adjustments

### PR DESCRIPTION
**Pull request by classicrocker883**
### Description
- the range did not allocate difference between Max and Min (+/-) and used just one value for both +/-  
also added `+` to make difference of +/-

- `LCD_BACKLIGHT_TIMEOUT_MINS`
added support for DWIN type displays for LCD timeout (in mins).  
LCD display and timeout and turn off after a certain amount of time has passed - screen saver-esque.

- in meshviewer, when it draws circles to indicate high or low spots on the bed, the circle colors have a more linear gradient - which is more true to how it should look, corresponding to values. also values aren't limited to 0.20. (fixed that).

- in the more recent commit for ProUI, `ADVANCE_PAUSE_FEATURE` "#if" statements were rearrange / ommited so that the adv pause feature can be disabled yet still utilize the other menu pages / functions.

- Updated CMakeLists.txt to stow warning for depreciated versions. Also added a "In case of..." / To Do If... comment.
in **WindowsPaths.cmake** or **UnixPaths.cmake** in the CMake install directory, an error may appear saying unknown command `_cmake_record_install_prefix()`. I give a fix which simply has the user either comment out (#) that line, or add `include(CMakeSystemSpecificInformation)` above it which links to the file in which that function is defined.

- Renamed PlotClass. Updated to fix spelling typo (*grph* => *graph*). Changed how graph is updated, in the graph each "block" amounts to (60/2) - **30** seconds of time rather than (50/2) - **25** seconds.
- Updated **proui/dwin.cpp** Variable update interval, from 1024 to 1000 (ms) and 500 (ms). which updates the menu a bit faster, temp - fan speed - blink - plot graph...

### Benefits
-stated in description-
for context, this is what the fix looks like
<p><a target="_blank" rel="noopener noreferrer nofollow"></a><p>
<br>
as you can see, the high and low values are correct. -0.24.....0.22+
otherwise originally it would read -0.24.....0.24+

<br>

**This also adds the ability to set a timeout for the LCD (PROUI), could be implemented in JyersUI and Creality_LCD.**

---
***For `SanityCheck`, it was edited to quiet errors, not sure if these changes are the best but it should be reviewed.

---

<br>

The fix to `dwinui.cpp` for `RainbowInt` -- better readability/implement code, draws the color gradient with wider spectrum (more true to color >> value)* .

The fix to `meshviewer.cpp` allows it to draw the actual true value, instead it was limited to |0.20| on the View Mesh.

also fixed [latest ProUI commit](https://github.com/MarlinFirmware/Marlin/commit/2a82ca87b03ab3b819f69b571ca44dd0a0e1afdf), rearrange ADVANCE_PAUSE_FEATURE functions so when it is disabled other functions like `parkHead` can be accessed.

---
most recent **plot.cpp plot.h** fixes typo, 30sec graph plot between white bars instead of 25sec.
remove redundant `DWIN_UPDATE_INTERVAL`  /  `DWIN_VAR_UPDATE_INTERVAL`, seperate the defined values to make better use. split 
`DWIN_VAR_UPDATE_INTERVAL` to update display variables in half time (faster).

### Configurations
`#define LCD_BACKLIGHT_TIMEOUT_MINS` can be uncommented - value of `10` for 10 minutes. 

### Related Issues
possible `SanityCheck` error compatibility with other configurations.

### CODE REVIEW

File: Marlin/src/lcd/e3v2/proui/meshviewer.cpp [hunk 1]
```diff
@@ -92,13 +92,13 @@ void MeshViewer::drawMeshPoint(const uint8_t x, const uint8_t y, const float z)
         DWINUI::drawSignedFloat(meshfont, 1, 1, px(x) - 3*fs, py(y) - fs, z);
         break;
       case -99 ... -1:
-        sprintf_P(str_1, PSTR("-.%02i"), -v);
+        sprintf_P(str_1, PSTR("%.2f"), z);
```
Comment (thinkyhead) [inline Marlin/src/lcd/e3v2/proui/meshviewer.cpp]: Sorry! `printf` on these platforms doesn't (always) support the `f` formatter. Use `MString` with `p_float_t`.

File: Marlin/src/lcd/marlinui.cpp [hunk 2]
```diff
@@ -190,11 +190,11 @@ constexpr uint8_t epps = ENCODER_PULSES_PER_STEP;
...
 
   void MarlinUI::refresh_backlight_timeout() {
-    backlight_off_ms = backlight_timeout_minutes ? millis() + backlight_timeout_minutes * 60UL * 1000UL : 0;
+    backlight_off_ms = backlight_timeout_minutes ? millis() + MIN_TO_MS(backlight_timeout_minutes) : 0;
     #ifdef NEOPIXEL_BKGD_INDEX_FIRST
       neo.reset_background_color();
       neo.show();
-    #elif PIN_EXISTS(LCD_BACKLIGHT)
+    #elif PIN_EXISTS(LCD_BACKLIGHT) && DISABLED(HAS_DWIN_E3V2)
       WRITE(LCD_BACKLIGHT_PIN, HIGH);
```
Comment (thinkyhead) [inline Marlin/src/lcd/marlinui.cpp]: Do we know for sure that none of the Ender3 V2 DWIN implementations want this?

**Comment by classicrocker883** [comment]
> Do we know for sure that none of the Ender3 V2 DWIN implementations want this? 

you are correct, you can omit ` && DISABLED(HAS_DWIN_E3V2)` on line 197. at the time it was something I had to amend, but now it seems to not be an issue as it was.

File: Marlin/src/lcd/e3v2/proui/bedlevel_tools.cpp [hunk 3]
```diff
@@ -252,22 +251,22 @@ bool BedLevelTools::meshValidate() {
...
-            sprintf_P(buf, PSTR("%s"), dtostrf(abs(bedlevel.z_values[x][y]), 1, 2, str_1));
+          MString<12> msg;
+          if ((GRID_MAX_POINTS_X) < TERN(TJC_DISPLAY, 8, 10))
+            msg.set(p_float_t(abs(bedlevel.z_values[x][y]), 2));
           else
-            sprintf_P(buf, PSTR("%02i"), (uint16_t)(abs(bedlevel.z_values[x][y] - (int16_t)bedlevel.z_values[x][y]) * 100));
-          offset_x = cell_width_px / 2 - (fs/2) * (strlen(buf)) - 2;
-          if (!(GRID_MAX_POINTS_X < (ENABLED(TJC_DISPLAY) ? 8 : 10)))
+            msg.setf_P(PSTR("%02i"), uint16_t(abs(bedlevel.z_values[x][y] - int16_t(bedlevel.z_values[x][y])) * 100));
+          offset_x = cell_width_px / 2 - (fs / 2) * strlen(buf) - 2;
```
Comment (classicrocker883) [inline Marlin/src/lcd/e3v2/proui/bedlevel_tools.cpp]: `strlen(buf)`

*buf* is undef, did you mean *msg*? 
 @thinkyhead

**Comment by classicrocker883** [comment]
for **/src/lcd/e3v2/proui/meshviewer.cpp**

can we just delete it so it is like
```y
 // if (sizex < (ENABLED(TJC_DISPLAY) ? 8 : 9)) {
 // if (v == 0) DWINUI::drawFloat(meshfont, 1, 2, px(x) - 2*fs, py(y) - fs, 0);
    if (v == 0) dwinDrawString(false, meshfont, DWINUI::textColor, DWINUI::backColor, px(x) - 2*fs, py(y) - fs, "0.00");
    else DWINUI::drawSignedFloat(meshfont, 1, 2, px(x) - 3*fs, py(y) - fs, z);
  /*}
  else {
    char str_1[9];
    str_1[0] = '\0';
    switch (v) {
      case -999 ... -100:
        DWINUI::drawSignedFloat(meshfont, 1, 1, px(x) - 3*fs, py(y) - fs, z);
        break;
      case -99 ... -1:
        sprintf_P(str_1, PSTR("%.2f"), z);
        break;
      case 0:
        dwinDrawString(false, meshfont, DWINUI::textColor, DWINUI::backColor, px(x) - 4, py(y) - fs, "0");
        break;
      case 1 ... 99:
        sprintf_P(str_1, PSTR("%.2f"), z);
        break;
      case 100 ... 999:
        DWINUI::drawSignedFloat(meshfont, 1, 1, px(x) - 3 * fs, py(y) - fs, z);
        break;
    }
    if (str_1[0])
      dwinDrawString(false, meshfont, DWINUI::textColor, DWINUI::backColor, px(x) - 2 * fs, py(y) - fs, str_1);
  }*/
}
```

am I missing something here? I mean does all that need to be there since `sprintf_P(str_1, PSTR(".%02i"), v);` does the same?

**Comment by classicrocker883** [comment]
otherwise would `msg.setf_P(PSTR("%02f"), z);` work?

```y
  else {
    char str_1[9];
    str_1[0] = 0;
    MString<12> msg;
    switch (v) {
      case -999 ... -100:  // -9.99 .. -1.00 mm
        DWINUI::drawSignedFloat(meshfont, 1, 1, px(x) - 3*fs, py(y) - fs, z);
        break;
      case -99 ... -1:  // -0.99 .. -0.01 mm
        msg.setf_P(PSTR("%02f"), z);
        //sprintf_P(str_1, PSTR("-.%02i"), -v);
        break;
      case 0:
        dwinDrawString(false, meshfont, DWINUI::textColor, DWINUI::backColor, px(x) - 4, py(y) - fs, "0");
        break;
      case 1 ... 99:  // 0.01 .. 0.99 mm
        msg.setf_P(PSTR("%02f"), z);
        //sprintf_P(str_1, PSTR(".%02i"), v);
        break;
      case 100 ... 999:  // 1.00 .. 9.99 mm
        DWINUI::drawSignedFloat(meshfont, 1, 1, px(x) - 3 * fs, py(y) - fs, z);
        break;
    }
    if (str_1[0])
      dwinDrawString(false, meshfont, DWINUI::textColor, DWINUI::backColor, px(x) - 2 * fs, py(y) - fs, msg); 
  }// REPLACED `str_1` with `msg`
```

or even  `msg.setf_P(PSTR("%f"), p_float_t(z, 2));`

**Comment by mriscoc** [comment]
Please @classicrocker883 stop copying code from my fork into the ProUI feature of Marlin. I want to decide when my code is ready to be ported.

**Comment by classicrocker883** [comment]
> Please @classicrocker883 stop copying code from my fork into the ProUI feature of Marlin. I want to decide when my code is ready to be ported, your changes will prevent me to continue to improve it in the future. 

this is already within the Marlin, i have not copied anything from "your" fork. relax. show me where I have copied from YOURS where you didnt copy from MINE???

what about all the code YOU copied from MINE???

its give and take, not take take take.

improving code is free. read the license

**Comment by classicrocker883** [comment]
> Please @classicrocker883 stop copying code from my fork into the ProUI feature of Marlin. I want to decide when my code is ready to be ported, your changes will prevent me to continue to improve it in the future. 

as you can see I posted this to the pull request over 2 weeks ago. before your most recent release.
https://github.com/MarlinFirmware/Marlin/pull/26086/commits/c282a4fdea241ae094db5ebb458d8a7d619129f9

I also have had this same code for a very long time. so dont tell me I am copying your code when you first copied from me.

thanks. have a good day.

**Comment by classicrocker883** [comment]
I think you are just very mistaken that this code was mine first to begin with. I reserve the right to alter any code that I had started before you copied from me. so dont say I copied from you first. thats all im saying. and we can go back to being friends.

like i said, anything i'm actually copying I would go by you first, that was my word.

and im not bypassing anything that originally was yours. so u can unblock me. again im sorry if you feel that way but I hope you realize that a group of words or phrase doesnt belong to a single person. if there solution to a code works there really is no other way of it being written. if you want to solve 2 + 2 no one owns that particular way of solving it or writing the code out.

@thinkyhead backme up here. i know youre busy, and sorry to drag you into this. but if you can mediate this and show who is right and who is wrong here?

**Comment by classicrocker883** [comment]
https://github.com/classicrocker883/MRiscoCProUI/commit/26ac60d9#diff-87d8e57accfb7dc24be7d8ea893999ab4c6a232eef87707a3399ebc34e564173

as you can see I first made these changes 10 months back in October last year.

I stand by I was not copying your code, I promise I wasnt trying to.

**Comment by classicrocker883** [comment]
there from ubl_G29.cpp I removed the comment you are so mad about. happy? did that make a difference?

>Ok, it is enough, I don't need to point line by line when you know what I am trying to say about your recent changes. I don't need to start a new battle. Good luck

did I really cause an issue for you in the future? it really upset me you claimed something I did which I didnt do. then you go and Block me?

do you honestly believe I still copied your code? clearly I had it first. Maybe you can see that and look past it.

if you forget this happened I will too.

the rest of the code is mine and therefore did not copy.

my function is `setTimer` which you dont have.

so much of this code is all mine which you had taken from me. its fine you used it, im proud you did, but dont go say I am copying your code which you know very well I did it first. ok?

**Comment by classicrocker883** [comment]
@mriscoc look here in **e3v2/proui/menus.cpp**
```y
    #if ENABLED(TITLE_CENTERED)
      DWINUI::drawCenteredString(false, DWIN_FONT_HEAD, hmiData.colorTitleTxt, hmiData.colorTitleBg, (TITLE_HEIGHT - DWINUI::fontHeight(DWIN_FONT_HEAD)) / 2 - 1, aTitle->caption);
    #else
      dwinDrawString(false, DWIN_FONT_HEAD, hmiData.colorTitleTxt, hmiData.colorTitleBg, 14, (TITLE_HEIGHT - DWINUI::fontHeight(DWIN_FONT_HEAD)) / 2 - 1, aTitle->caption);
    #endif
```

this code change was originally done by me.  
`TITLE_CENTERED`?? you have that no where else in your code.  
this is just one of many examples where you taken from me.

yet you wont let this go, I mean seriously how does my pull request effect you?

I really think you want all the credit for yourself, otherwise why would you block me?

**Comment by mriscoc** [comment]
> this code change was originally done by me.
> `TITLE_CENTERED`?? you have that no where else in your code.
> this is just one of many examples where you taken from me.

Please stop, I made the possibility to center the title because you request it, I didn't agree but thought it might be a harmless change. Your credit is here: https://github.com/mriscoc/Ender3V2S1/issues/270

**Comment by classicrocker883** [comment]
no forget it. we're good don't worry about it

File: Marlin/src/lcd/e3v2/proui/bedlevel_tools.cpp (hunk 3 above)
Comment (classicrocker883) [reply Marlin/src/lcd/e3v2/proui/bedlevel_tools.cpp]: jyersui has this line like so

```y
  msg.setf(F("%02i"), uint16_t(abs(bedlevel.z_values[x][y] - int16_t(bedlevel.z_values[x][y])) * 100));
offset_x = cell_width_px / 2 - 3 * msg.length() - 2;
```

**Comment by thinkyhead** [comment]
> would `msg.setf_P(PSTR("%02f"), z)` work?

Keep in mind that we cannot use `%f` with these platforms. To print floats we must use `dtostrf`, which is handled by the `MString` class using `float` — or one of the wrappers, `p_float_t` or `w_float_t` to set a specific precision.

**Comment by classicrocker883** [comment]
> > would `msg.setf_P(PSTR("%02f"), z)` work?
> 
> Keep in mind that we cannot use `%f` with these platforms. To print floats we must use `dtostrf`, which is handled by the `MString` class using `float` — or one of the wrappers, `p_float_t` or `w_float_t` to set a specific precision.

**Comment by classicrocker883** [comment]
everything you changed is looking good. however a couple things needed to addressed, I did the necessary tweaks. had to omit and revert back the range for JyersUI/**dwin.cpp** line ~400. if say you went ahead and used `rmin` with `rmax` to define the color output for values, you don't get the expected gradient.

in this part of the function  `void drawBedMesh`
```y
      // Draw value square grid
      GRID_LOOP(x, y) {
```

I know this because I tested it that way, having `rmin` and `rmax` (as `range` and `range2`) it just doesn't work well. like the highest value should be the brightest color, however this way makes it not so, and another value below that will be brighter - so it gets very confusing.

**Comment by thinkyhead** [comment]
I went ahead and cleaned up the `gcode_preview.*` code to address some things.
- Move `Preview_` functions into a static singleton class.
- Define `fileprop_t` methods inside the struct.
- Apply `PSTR` in case this code is ever deployed on AVR.
- Return `bool` types in `bool` methods.
- Use `p_float_t` instead of `dtostrf` for some strings.
- Move `thumbdata` from the heap to the stack (assuming the stack can be very large).

**Comment by classicrocker883** [comment]
> I went ahead and cleaned up the `gcode_preview.*` code to address some things.


thanks for taking the time and doing that, its looking good, ill give it a test shot. quick question regarding the G-code preview feature, when selecting to view a file with a thumbnail more than once, following these steps:

1. Go to Print menu
2. Select any file w/thumbnail preview
3. Cancel/return to file menu
4. Repeat Step 2 and 3

depending on the file, you can do this twice or sometimes 3 or 4 times until it freezes/crashes and reboots.  
however if you say go to another folder or select another file *w/out* a thumbnail preview, this sort of clears the cache and you can do this as many times as you like.

so obviously somewhere in memory where the image is stored it overloads, there should be a way to clear the cache or address when selecting **Cancel** instead of print. I've tried a few different things that haven't worked like `delete[] fileprop.thumbnail`

more about it here. https://github.com/MarlinFirmware/Marlin/issues/26169

while we are on the subject is there anything which may alleviate the situation?

**Comment by classicrocker883** [comment]
just wanted to comment saying I updated the **plot.cpp plot.h** and made `DWIN_VAR_UPDATE_INTERVAL` useful in **proui/dwin.cpp**

**Comment by thinkyhead** [comment]
> while we are on the subject is there anything which may alleviate the situation?

I found one possible culprit. The `buf64[]` array allocates `fileprop.thumbsize` bytes, then loops until `nread == fileprop.thumbsize`, and then `buf64[nread] = '\0'`. That last step exceeds the buffer size. So I just patched that.

If this doesn't fix the problem then we should look further.

**Comment by mriscoc** [comment]
Good PR! @classicrocker883 thanks for your contributions, feel free to port the features that you consider helpful from my fork to Marlin, I think that this will help me henceforth to fully focus in my fork.

**Pull request merged**